File: facecat_cpp/source/input/FCComboBox.cpp
#include "..\\..\\stdafx.h"
#include "..\\..\\include\\input\\FCComboBox.h"

namespace FaceCat{
	FCComboBoxMenu::FCComboBoxMenu(){
		m_comboBox = 0;
	}

	FCComboBoxMenu::~FCComboBoxMenu(){
		m_comboBox = 0;
	}

	FCComboBox* FCComboBoxMenu::getComboBox(){
		return m_comboBox;
	}

	void FCComboBoxMenu::setComboBox(FCComboBox *comboBox){
		m_comboBox = comboBox;
	}

	bool FCComboBoxMenu::onAutoHide(){
		if(m_comboBox && m_comboBox->isFocused()){
			return false;
		}
		return true;
	}

	///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	void FCComboBox::dropDownButtonTouchDown(Object sender, FCTouchInfo touchInfo, Object pInvoke){
		FCComboBox *comboBox = (FCComboBox*)pInvoke;
		if(comboBox){
			comboBox->onDropDownOpening();
		}
	}

	void FCComboBox::menuItemClick(Object sender, FCMenuItem *item, FCTouchInfo touchInfo, Object pInvoke){
		FCComboBox *comboBox = (FCComboBox*)pInvoke;
		comboBox->setText(item->getText());
		ArrayList<FCMenuItem*> items = comboBox->getItems();
        int itemSize = (int)items.size();
        for (int i = 0; i < itemSize; i++){
            if (items.get(i) == item){
                comboBox->setSelectedIndex(i);
                break;
            }
        }
		comboBox->setSelectionStart((int)comboBox->getText().length());
		comboBox->invalidate();
	}

	void FCComboBox::menuKeyDown(Object sender, char key, Object pInvoke){
		if(key == 13){
			FCComboBox *comboBox = (FCComboBox*)pInvoke;
			if(comboBox){
				comboBox->onSelectedIndexChanged();
			}
		}
	}

	///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	FCComboBox::FCComboBox(){
		m_dropDownButton = 0;
		m_dropDownButtonTouchDownEvent = dropDownButtonTouchDown;
		m_dropDownMenu = 0;
		m_menuItemClickEvent = menuItemClick;
		m_menuKeyDownEvent = menuKeyDown;
	}

	FCComboBox::~FCComboBox(){
		if(m_dropDownButton){
			if(m_dropDownButtonTouchDownEvent){
				m_dropDownButton->removeEvent(m_dropDownButtonTouchDownEvent, FCEventID::TOUCHDOWN);
				m_dropDownButtonTouchDownEvent = 0;
			}
			m_dropDownButton = 0;
		}
		if(m_dropDownMenu){
			if(m_menuItemClickEvent){
				m_dropDownMenu->removeEvent(m_menuItemClickEvent, FCEventID::MENUITEMCLICK);
				m_menuItemClickEvent = 0;
			}
			if(m_menuKeyDownEvent){
				m_dropDownMenu->removeEvent(m_menuKeyDownEvent, FCEventID::KEYDOWN);
				m_menuKeyDownEvent = 0;
			}
			getNative()->removeControl(m_dropDownMenu);
			delete m_dropDownMenu;
			m_dropDownMenu = 0;
		}
	}

	FCButton* FCComboBox::getDropDownButton(){
		return m_dropDownButton;
	}

	FCComboBoxMenu* FCComboBox::getDropDownMenu(){
		return m_dropDownMenu;
	}

	int FCComboBox::getSelectedIndex(){
		if (m_dropDownMenu){
			ArrayList<FCMenuItem*> items = m_dropDownMenu->m_items;
			int itemSize = (int)items.size();
			for (int i = 0; i < itemSize; i++){
				FCMenuItem *item = items.get(i);
				if (item->isChecked()){
					return i;
				}
			}
		}
		return -1;
	}

	void FCComboBox::setSelectedIndex(int selectedIndex){
		if (m_dropDownMenu){
			ArrayList<FCMenuItem*> items = m_dropDownMenu->m_items;
			int itemSize = (int)items.size();
			bool changed = false;
			for (int i = 0; i < itemSize; i++){
				FCMenuItem *item = items.get(i);
				if (i == selectedIndex){
					 if (!item->isChecked()){
						 item->setChecked(true);
						 changed = true;
					 }
					 setText(item->getText());
				 }
				 else{
					 item->setChecked(false);
				 }
			 }
			if(changed){
				onSelectedIndexChanged();
			}
		}
	}

	String FCComboBox::getSelectedText(){
		if (m_dropDownMenu){
			ArrayList<FCMenuItem*> items = m_dropDownMenu->m_items;
			int itemSize = (int)items.size();
			for (int i = 0; i < itemSize; i++){
				FCMenuItem *item = items.get(i);
				if (item->isChecked()){
					return item->getText();
				}
			}
		}
		return L"";
	}

	void FCComboBox::setSelectedText(const String& selectedText){
		if (m_dropDownMenu){
			ArrayList<FCMenuItem*> items = m_dropDownMenu->m_items;
			int itemSize = (int)items.size();
			bool changed = false;
			for (int i = 0; i < itemSize; i++){
				FCMenuItem *item = items.get(i);
				if (item->getText() == selectedText){
					 if (!item->isChecked()){
						 item->setChecked(true);
						 changed = true;
					 }
					 setText(item->getText());
				 }
				 else{
					 item->setChecked(false);
				 }
			 }
			if(changed){
				onSelectedIndexChanged();
			}
		}
	}

	String FCComboBox::getSelectedValue(){
		if (m_dropDownMenu){
			ArrayList<FCMenuItem*> items = m_dropDownMenu->m_items;
			int itemSize = (int)items.size();
			for (int i = 0; i < itemSize; i++){
				FCMenuItem *item = items.get(i);
				if (item->isChecked()){
					return item->getValue();
				}
			}
		}
		return L"";
	}

	void FCComboBox::setSelectedValue(const String& selectedValue){
		if (m_dropDownMenu){
			ArrayList<FCMenuItem*> items = m_dropDownMenu->m_items;
			int itemSize = (int)items.size();
			bool changed = false;
			for (int i = 0; i < itemSize; i++){
				FCMenuItem *item = items.get(i);
				if (item->getValue() == selectedValue){
					 if (!item->isChecked()){
						 item->setChecked(true);
						 changed = true;
					 }
					 setText(item->getText());
				 }
				 else{
					 item->setChecked(false);
				 }
			 }
			if(changed){
				onSelectedIndexChanged();
			}
		}
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	void FCComboBox::addItem(FCMenuItem *item){
		if(m_dropDownMenu){
			m_dropDownMenu->addItem(item);
		}
	}

	void FCComboBox::clearItems(){
		if(m_dropDownMenu){
			m_dropDownMenu->clearItems();
		}
	}

	String FCComboBox::getControlType(){
		return L"ComboBox";
	}

	ArrayList<FCMenuItem*> FCComboBox::getItems(){
		ArrayList<FCMenuItem*> items;
		if(m_dropDownMenu){
			items = m_dropDownMenu->m_items;
		}
		return items;
	}

	void FCComboBox::getProperty(const String& name, String *value, String *type){
		if(name == L"selectedindex"){
			*type = L"int";
			*value = FCStr::convertIntToStr(getSelectedIndex());
		}
		else if(name == L"selectedtext"){
			*type = L"text";
			*value = getSelectedText();
		}
		else if(name == L"selectedvalue"){
			*type = L"text";
			*value = getSelectedValue();
		}
		else{
			FCTextBox::getProperty(name, value, type);
		}
	}

	ArrayList<String> FCComboBox::getPropertyNames(){
		ArrayList<String> propertyNames = FCTextBox::getPropertyNames();
		propertyNames.add(L"SelectedIndex");
		propertyNames.add(L"SelectedText");
		propertyNames.add(L"SelectedValue");
		return propertyNames;
	}

	void FCComboBox::insertItem(int index, FCMenuItem *item){
		if(m_dropDownMenu){
			m_dropDownMenu->insertItem(index, item);
		}
	}

	void FCComboBox::onDropDownOpening(){
		FCComboBoxMenu *menu = getDropDownMenu();
		menu->setNative(getNative());
		FCPoint popupPoint = {0, getHeight()};
		FCPoint nativePoint = pointToNative(popupPoint);
		menu->setLocation(nativePoint);
		FCSize mSize = {getWidth(), menu->getContentHeight()};
		menu->setSize(mSize);
		menu->setVisible(true);
		menu->bringToFront();
		menu->invalidate();
	}

	void FCComboBox::onKeyDown(char key){
		FCTextBox::onKeyDown(key);
		FCHost *host = getNative()->getHost();
		if(!host->isKeyPress(VK_CONTROL)
		&& !host->isKeyPress(VK_MENU)
		&& !host->isKeyPress(VK_SHIFT)){
			if(getLinesCount() <= 1){
				if(key == 13 || key == 38 || key == 40){
					if(m_dropDownMenu){
						m_dropDownMenu->onKeyDown(key);
					}
				}
			}
		}
	}
	
	void FCComboBox::onLoad(){
		FCTextBox::onLoad();
		FCHost *host = getNative()->getHost();
		if (!m_dropDownButton){
			m_dropDownButton = dynamic_cast<FCButton*>(host->createInternalControl(this, L"dropdownbutton"));
			addControl(m_dropDownButton);
			m_dropDownButton->addEvent(m_dropDownButtonTouchDownEvent, FCEventID::TOUCHDOWN, this);
		}
		if (!m_dropDownMenu){
			m_dropDownMenu = dynamic_cast<FCComboBoxMenu*>(host->createInternalControl(this, L"dropdownmenu"));
			getNative()->addControl(m_dropDownMenu);
			m_dropDownMenu->setVisible(false);
			m_dropDownMenu->addEvent(m_menuItemClickEvent, FCEventID::MENUITEMCLICK, this);
			m_dropDownMenu->addEvent(m_menuKeyDownEvent, FCEventID::KEYDOWN, this);
		}
		else{
			m_dropDownMenu->setNative(getNative());
		}
	}

	void FCComboBox::onSelectedIndexChanged(){
		callEvents(FCEventID::SELECTEDINDEXCHANGED);
	}

	void FCComboBox::onTouchWheel(FCTouchInfo touchInfo){
		FCTextBox::onTouchWheel(touchInfo);
		if(getLinesCount() <= 1){
			if(m_dropDownMenu){
				FCTouchInfo newTouchInfo = touchInfo;
                newTouchInfo.m_firstPoint = m_dropDownMenu->getTouchPoint();
                newTouchInfo.m_secondPoint = m_dropDownMenu->getTouchPoint();
				m_dropDownMenu->onTouchWheel(newTouchInfo);
			}
		}
	}

	void FCComboBox::removeItem(FCMenuItem *item){
		if(m_dropDownMenu){
			m_dropDownMenu->removeItem(item);
		}
	}

	void FCComboBox::setProperty(const String& name, const String& value){
		if(name == L"selectedindex"){
			setSelectedIndex(FCStr::convertStrToInt(value));
		}
		else if(name == L"selectedtext"){
			setSelectedText(value);
		}
		else if(name == L"selectedvalue"){
			setSelectedValue(value);
		}
		else{
			FCTextBox::setProperty(name, value);
		}
	}

	void FCComboBox::update(){
		int width = getWidth(), height = getHeight(), uBottom = 0;
		if (m_dropDownButton){
			int dWidth = m_dropDownButton->getWidth();
			FCPoint location = {width - dWidth, 0};
			m_dropDownButton->setLocation(location);
			FCSize size = {dWidth, height};
			m_dropDownButton->setSize(size);
			FCPadding padding(0, 0, dWidth, 0);
			setPadding(padding);
		}
	}
}

File: facecat_java/src/facecat/topin/input/FCComboBox.java
/*捂脸猫FaceCat框架 v1.0
 1.创始人-矿洞程序员-上海宁米科技创始人-脉脉KOL-陶德 (微信号:suade1984);
 2.联合创始人-上海宁米科技创始人-袁立涛(微信号:wx627378127);
 3.联合创始人-河北思尔企业管理咨询有限公司合伙人-肖添龙(微信号:xiaotianlong_luu);
 4.联合开发者-陈晓阳(微信号:chenxiaoyangzxy)，助理-朱炜(微信号:cnnic_zhu);
 5.该框架开源协议为BSD，欢迎对我们的创业活动进行各种支持，欢迎更多开发者加入。
 包含C/C++,Java,C#,iOS,MacOS,Linux六个版本的图形和通讯服务框架。
 */
package facecat.topin.input;

import facecat.topin.div.*;
import facecat.topin.btn.*;
import facecat.topin.core.*;
import java.util.*;

/**
 * 下拉列表控件
 */
public class FCComboBox extends FCTextBox implements FCMenuItemTouchEvent, FCTouchEvent {

    /**
     * 创建下拉列表
     */
    public FCComboBox() {
    }

    protected FCButton m_dropDownButton = null;

    /**
     * 获取下拉按钮
     */
    public FCButton getDropDownButton() {
        return m_dropDownButton;
    }

    protected FCComboBoxMenu m_dropDownMenu = null;

    /**
     * 获取下拉菜单
     */
    public FCComboBoxMenu getDropDownMenu() {
        return m_dropDownMenu;
    }

    /**
     * 获取选中的索引
     */
    public int getSelectedIndex() {
        if (m_dropDownMenu != null) {
            ArrayList<FCMenuItem> items = m_dropDownMenu.getItems();
            int itemSize = items.size();
            for (int i = 0; i < itemSize; i++) {
                FCMenuItem item = items.get(i);
                if (item.isChecked()) {
                    return i;
                }
            }
        }
        return -1;
    }

    /**
     * 设置选中的索引
     */
    public void setSelectedIndex(int value) {
        if (m_dropDownMenu != null) {
            ArrayList<FCMenuItem> items = m_dropDownMenu.getItems();
            int itemSize = items.size();
            boolean changed = false;
            for (int i = 0; i < itemSize; i++) {
                FCMenuItem item = items.get(i);
                if (i == value) {
                    if (!item.isChecked()) {
                        // 选中并设置文字
                        item.setChecked(true);
                        changed = true;
                    }
                    setText(item.getText());
                } else {
                    item.setChecked(false);
                }
            }
            if (changed) {
                onSelectedIndexChanged();
            }
        }
    }

    /**
     * 获取选中的文字
     */
    public String getSelectedText() {
        if (m_dropDownMenu != null) {
            ArrayList<FCMenuItem> items = m_dropDownMenu.getItems();
            int itemSize = items.size();
            for (int i = 0; i < itemSize; i++) {
                FCMenuItem item = items.get(i);
                if (item.isChecked()) {
                    return item.getText();
                }
            }
        }
        return null;
    }

    /**
     * 设置选中的文字
     */
    public void setSelectedText(String value) {
        if (m_dropDownMenu != null) {
            ArrayList<FCMenuItem> items = m_dropDownMenu.getItems();
            int itemSize = items.size();
            boolean changed = false;
            for (int i = 0; i < itemSize; i++) {
                FCMenuItem item = items.get(i);
                if (item.getText().equals(value)) {
                    // 选中并设置文字
                    if (!item.isChecked()) {
                        item.setChecked(true);
                        changed = true;
                    }
                    setText(item.getText());
                } else {
                    item.setChecked(false);
                }
            }
            if (changed) {
                onSelectedIndexChanged();
            }
        }
    }

    /**
     * 获取选中的值
     */
    public String getSelectedValue() {
        if (m_dropDownMenu != null) {
            ArrayList<FCMenuItem> items = m_dropDownMenu.getItems();
            int itemSize = items.size();
            for (int i = 0; i < itemSize; i++) {
                FCMenuItem item = items.get(i);
                if (item.isChecked()) {
                    return item.getValue();
                }
            }
        }
        return null;
    }

    /**
     * 设置选中的值
     */
    public void setSelectedValue(String value) {
        if (m_dropDownMenu != null) {
            ArrayList<FCMenuItem> items = m_dropDownMenu.getItems();
            int itemSize = items.size();
            boolean changed = false;
            for (int i = 0; i < itemSize; i++) {
                FCMenuItem item = items.get(i);
                if (item.getValue().equals(value)) {
                    // 选中并设置文字
                    if (!item.isChecked()) {
                        item.setChecked(true);
                        changed = true;
                    }
                    setText(item.getText());
                } else {
                    item.setChecked(false);
                }
            }
            if (changed) {
                onSelectedIndexChanged();
            }
        }
    }

    /**
     * 添加菜单项
     *
     * @param item 菜单项
     */
    public void addItem(FCMenuItem item) {
        if (m_dropDownMenu != null) {
            m_dropDownMenu.addItem(item);
        }
    }

    @Override
    public void callControlTouchEvent(int eventID, Object sender, FCTouchInfo touchInfo) {
        if (sender == m_dropDownButton) {
            if (eventID == FCEventID.TOUCHDOWN) {
                onDropDownOpening();
            }
        }
    }

    public void callMenuItemTouchEvent(int eventID, Object sender, FCMenuItem item, FCTouchInfo touchInfo) {
        if (sender == m_dropDownMenu) {
            if (eventID == FCEventID.MENUITEMCLICK) {
                setText(item.getText());
                ArrayList<FCMenuItem> items = getItems();
                int itemSize = items.size();
                for (int i = 0; i < itemSize; i++) {
                    if (items.get(i) == item) {
                        setSelectedIndex(i);
                        break;
                    }
                }
                setSelectionStart(getText().length());
                invalidate();
            }
        }
    }

    /**
     * 销毁方法
     */
    @Override
    public void delete() {
        if (!isDeleted()) {
            if (m_dropDownButton != null) {
                m_dropDownButton.removeEvent(this, FCEventID.TOUCHDOWN);
            }
            if (m_dropDownMenu != null) {
                m_dropDownMenu.removeEvent(this, FCEventID.MENUITEMCLICK);
                getNative().removeControl(m_dropDownMenu);
                m_dropDownMenu.delete();
                m_dropDownMenu = null;
            }
        }
        super.delete();
    }

    /**
     * 清除所有菜单项
     */
    public void clearItems() {
        if (m_dropDownMenu != null) {
            m_dropDownMenu.clearItems();
        }
    }

    /**
     * 获取控件类型
     */
    @Override
    public String getControlType() {
        return "ComboBox";
    }

    /**
     * 获取菜单项
     */
    public ArrayList<FCMenuItem> getItems() {
        if (m_dropDownMenu != null) {
            return m_dropDownMenu.getItems();
        }
        return null;
    }

    /**
     * 获取属性值
     *
     * @param name 属性名称
     * @param value 返回属性值
     * @param type 返回属性类型
     */
    @Override
    public void getProperty(String name, RefObject<String> value, RefObject<String> type) {
        if (name.equals("selectedindex")) {
            type.argvalue = "int";
            value.argvalue = FCStr.convertIntToStr(getSelectedIndex());
        } else if (name.equals("selectedtext")) {
            type.argvalue = "text";
            value.argvalue = getSelectedText();
        } else if (name.equals("selectedvalue")) {
            type.argvalue = "text";
            value.argvalue = getSelectedValue();
        } else {
            super.getProperty(name, value, type);
        }
    }

    /**
     * 获取属性名称列表
     */
    @Override
    public ArrayList<String> getPropertyNames() {
        ArrayList<String> propertyNames = super.getPropertyNames();
        propertyNames.addAll(Arrays.asList(new String[]{"SelectedIndex", "SelectedText", "SelectedValue"}));
        return propertyNames;
    }

    /**
     * 插入菜单项
     *
     * @param index 索引
     * @param item 菜单项
     */
    public void insertItem(int index, FCMenuItem item) {
        if (m_dropDownMenu != null) {
            m_dropDownMenu.insertItem(index, item);
        }
    }

    /**
     * 菜单下拉方法
     */
    public void onDropDownOpening() {
        m_dropDownMenu.setNative(getNative());
        FCPoint nativePoint = pointToNative(new FCPoint(0, getHeight()));
        m_dropDownMenu.setLocation(nativePoint);
        m_dropDownMenu.setSize(new FCSize(getWidth(), m_dropDownMenu.getContentHeight()));
        m_dropDownMenu.setVisible(true);
        m_dropDownMenu.bringToFront();
        m_dropDownMenu.invalidate();
    }

    /**
     * 添加控件方法
     */
    @Override
    public void onLoad() {
        super.onLoad();
        FCHost host = getNative().getHost();
        if (m_dropDownButton == null) {
            FCView tempVar = host.createInternalControl(this, "dropdownbutton");
            m_dropDownButton = (FCButton) ((tempVar instanceof FCButton) ? tempVar : null);
            addControl(m_dropDownButton);
            m_dropDownButton.addEvent(this, FCEventID.TOUCHDOWN);
        }
        if (m_dropDownMenu == null) {
            FCView tempVar2 = host.createInternalControl(this, "dropdownmenu");
            m_dropDownMenu = (FCComboBoxMenu) ((tempVar2 instanceof FCComboBoxMenu) ? tempVar2 : null);
            getNative().addControl(m_dropDownMenu);
            m_dropDownMenu.setVisible(false);
            m_dropDownMenu.addEvent(this, FCEventID.MENUITEMCLICK);
        } else {
            m_dropDownMenu.setNative(getNative());
        }
    }

    /**
     * 选中索引改变方法
     */
    public void onSelectedIndexChanged() {
        callEvents(FCEventID.SELECTEDINDEXCHANGED);
    }

    /**
     * 移除菜单项
     *
     * @param item 菜单项
     */
    public void removeItem(FCMenuItem item) {
        if (m_dropDownMenu != null) {
            m_dropDownMenu.removeItem(item);
        }
    }

    /**
     * 设置属性值
     *
     * @param name 属性名称
     * @param value 属性值
     */
    @Override
    public void setProperty(String name, String value) {
        if (name.equals("selectedindex")) {
            setSelectedIndex(FCStr.convertStrToInt(value));
        } else if (name.equals("selectedtext")) {
            setSelectedText(value);
        } else if (name.equals("selectedvalue")) {
            setSelectedValue(value);
        } else {
            super.setProperty(name, value);
        }
    }

    /**
     * 更新布局方法
     */
    @Override
    public void update() {
        super.update();
        int width = getWidth(), height = getHeight();
        if (m_dropDownButton != null) {
            int dWidth = m_dropDownButton.getWidth();
            m_dropDownButton.setLocation(new FCPoint(width - dWidth, 0));
            m_dropDownButton.setSize(new FCSize(dWidth, height));
            setPadding(new FCPadding(0, 0, dWidth, 0));
        }
    }
}
